tests/GraphQL/AddElementToAreaMutationTest.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+namespace DNADesign\Elemental\Tests\GraphQL;
+
+use DNADesign\Elemental\GraphQL\AddElementToAreaMutation;
+use DNADesign\Elemental\Models\ElementalArea;
+use DNADesign\Elemental\Tests\Src\TestElement;
+use GraphQL\Type\Definition\ResolveInfo;
+use InvalidArgumentException;
+use SilverStripe\Dev\SapphireTest;
+use SilverStripe\Security\Security;
+
+class AddElementToAreaMutationTest extends SapphireTest
+{
+    protected static $fixture_file = 'AddElementToAreaMutationTest.yml';
+
+    protected static $extra_dataobjects = [
+        TestElement::class,
+    ];
+
+    public function testAddingBlocksInOrder()
+    {
+        $className = TestElement::class;
+
+        $this->runMutation($className, 1);
+        $this->assertIdsInOrder([1], 'The first element is added');
+
+        $this->runMutation($className, 1, 1);
+        $this->assertIdsInOrder([1, 2], 'The second element is added after the first');
+
+        $this->runMutation($className, 1, 1);
+        $this->assertIdsInOrder([1, 3, 2], 'The third element is added after the first');
+
+        $this->runMutation($className, 1);
+        $this->assertIdsInOrder([1, 3, 2, 4], 'The fourth element is added last, when no after ID parameter is given');
+
+        $this->runMutation($className, 1, 0);
+        $this->assertIdsInOrder([5, 1, 3, 2, 4], 'The fifth element is added first, after ID 0');
+    }
+
+    /**
+     * @expectedException InvalidArgumentException
+     */
+    public function testBadElementalArea()
+    {
+        $this->runMutation(TestElement::class, 2);
+    }
+
+    /**
+     * @expectedException InvalidArgumentException
+     */
+    public function testOrderingByWrongElementalArea()
+    {
+        $firstArea = ElementalArea::get()->first();
+        $elementInFirstArea = TestElement::create();
+        $firstArea->Elements()->add($elementInFirstArea);
+
+        ElementalArea::create()->write();
+        $this->runMutation(TestElement::class, 2, $elementInFirstArea->ID);
+    }
+
+    protected function assertIdsInOrder($ids, $message = null)
+    {
+        $actualIDs = TestElement::get()->sort('Sort')->map()->keys();
+
+        $this->assertSame($ids, $actualIDs, $message);
+    }
+
+    protected function runMutation($className, $elementalAreaID, $afterElementId = null)
+    {
+        $mutation = new AddElementToAreaMutation();
+        $context = ['currentUser' => Security::getCurrentUser()];
+        $resolveInfo = new ResolveInfo([]);
+
+        $args = [
+            'ClassName' => $className,
+            'ElementalAreaID' => $elementalAreaID,
+        ];
+
+        if (!is_null($afterElementId)) {
+            $args['AfterElementID'] = $afterElementId;
+        }
+
+        $mutation->resolve(null, $args, $context, $resolveInfo);
```

Comment:
You will probably have to have a way of returning IDs here so that you can compare against them in your continued tests. Alternatively you can seed with a fixture and specify IDs. You can't 100% rely on IDs being ascending order from 1 in functional tests.

Reply:
Sorry I don't understand what you're trying to tell me.

Reply:
You add a block with the mutation, then assume the ID of that block is `1`. It would be best practice to get the ID from the result of the `resolve` method and use that as the source of truth for the ID rather than assuming it's going to be `1`.

Reply:
FYI this is why the "SortBlockMutation" test fixture specified IDs.

Reply:
Yeah I think I agree. Using hard coded IDs in tests is bad and introduces unnecessary tech debt that we'll have to fix in future. FYI there's discussions in core at the moment about abstracting the primary key assignment for models so UUIDs would become the default - pieces of code like this are the main reason we're talking about not doing it in 4.x at the moment because it'll break people's tests (potentially)

Reply:
FYI we do set IDs in the fixtures for SortBlocks. I don't think seeding the IDs in that manor is particularly bad? Even if the IDs switch to UUID manually setting the IDs to int would probably still work?

Reply:
I guess the better option is to assert orders based on some piece of fixed data like a model's Title or whatever

Reply:
OK, I can do this.

However I'd like to also point out that while I can agree that one should code to an abstraction rather than an implementation, in order to switch the IDs out to e.g. UUID there would need to be much more work than simply altering these tests - all ID's are currently typed as Integers, and GraphQL is strongly typed. It is not simply tests that would break.

Reply:
I think this conversation needs to head over to the RFC https://github.com/silverstripe/silverstripe-framework/issues/8411